File: ml.Rcheck/tests/testthat/test-shelf.R
test_that("shelf returns ml_shelf_result", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_s3_class(result, "ml_shelf_result")
})

test_that("shelf $fresh is logical", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_true(is.logical(result$fresh))
})

test_that("shelf $stale is inverse of $fresh", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_equal(result$stale, !result$fresh)
})

test_that("shelf $metrics_then populated from model scores_", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_true(length(result$metrics_then) > 0L)
})

test_that("shelf $metrics_now populated from evaluate()", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_true(length(result$metrics_now) > 0L)
})

test_that("shelf $degradation is a named list", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_true(is.list(result$degradation))
})

test_that("shelf model with no scores_ raises model_error", {
  s     <- iris_split()
  model <- ml_fit(s$train, "Species", algorithm = "logistic", seed = 42L)
  # holdout fit → no scores_
  expect_error(
    ml_shelf(model, new = iris[1:30, ], target = "Species"),
    class = "model_error"
  )
})

test_that("shelf target mismatch raises config_error", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  expect_error(
    ml_shelf(model, new = iris[1:30, ], target = "Sepal.Length"),
    class = "config_error"
  )
})

test_that("shelf error when new data has < 5 rows", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  expect_error(
    ml_shelf(model, new = iris[1:3, ], target = "Species"),
    class = "data_error"
  )
})

test_that("shelf auto-unwraps TuningResult", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  tuned <- ml_tune(cv, "Species", algorithm = "logistic",
                   n_trials = 2L, seed = 42L)
  # TuningResult best_model may have scores_ if CV was used
  # Just test it doesn't crash on the unwrapping path
  tryCatch(
    ml_shelf(tuned, new = iris[1:30, ], target = "Species"),
    error = function(e) {
      # model_error is OK if no scores_ — just check it's not a class error
      expect_false(inherits(e, "simpleError") && grepl("ml_tuning_result", conditionMessage(e)))
    }
  )
})

test_that("shelf $recommendation is a non-empty string", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_true(is.character(result$recommendation))
  expect_true(nchar(result$recommendation) > 0L)
})

test_that("shelf fresh=TRUE with very high tolerance", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species", tolerance = 1.0)
  expect_true(result$fresh)
})

test_that("shelf print works", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml_shelf(model, new = iris[1:30, ], target = "Species")
  expect_output(print(result), "Shelf")
})

test_that("ml$shelf() module style works", {
  cv    <- ml_split(iris, "Species", seed = 42L, folds = 3L)
  model <- ml_fit(cv, "Species", algorithm = "logistic", seed = 42L)
  result <- ml$shelf(model, new = iris[1:30, ], target = "Species")
  expect_s3_class(result, "ml_shelf_result")
})

File: python/tests/test_shelf.py
"""Tests for ml.shelf() — label-required model freshness check."""

import numpy as np
import pandas as pd
import pytest

import ml
from ml._types import ConfigError, DataError
from ml.shelf import ShelfResult, shelf

# ── Fixtures ──────────────────────────────────────────────────────────────────

@pytest.fixture
def clf_df():
    rng = np.random.RandomState(42)
    n = 300
    x1 = rng.rand(n)
    return pd.DataFrame({
        "x1": x1,
        "x2": rng.rand(n),
        "target": (x1 > 0.5).astype(str),  # separable signal
    })


@pytest.fixture
def clf_split(clf_df):
    return ml.split(data=clf_df, target="target", seed=42)


@pytest.fixture
def clf_model(clf_df):
    """CV-fitted model so scores_ is populated (required for shelf comparison)."""
    s = ml.split(data=clf_df, target="target", seed=42)
    cv = ml.cv(s, folds=2, seed=42)
    return ml.fit(data=cv, target="target", seed=42)


@pytest.fixture
def reg_df():
    rng = np.random.RandomState(7)
    n = 300
    return pd.DataFrame({
        "x1": rng.rand(n),
        "x2": rng.rand(n),
        "target": rng.rand(n),
    })


@pytest.fixture
def reg_split(reg_df):
    return ml.split(data=reg_df, target="target", seed=42)


@pytest.fixture
def reg_model(reg_df):
    s = ml.split(data=reg_df, target="target", seed=42)
    cv = ml.cv(s, folds=2, seed=42)
    return ml.fit(data=cv, target="target", seed=42)


# ── Return type ───────────────────────────────────────────────────────────────

def test_shelf_returns_shelfresult(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result, ShelfResult)


def test_shelf_via_ml_namespace(clf_model, clf_split):
    result = ml.shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result, ShelfResult)


# ── Attributes ────────────────────────────────────────────────────────────────

def test_shelf_has_fresh(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result.fresh, bool)


def test_shelf_has_metrics_then(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result.metrics_then, dict)
    assert len(result.metrics_then) > 0


def test_shelf_has_metrics_now(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result.metrics_now, dict)
    assert len(result.metrics_now) > 0


def test_shelf_has_degradation(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result.degradation, dict)


def test_shelf_has_recommendation(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert isinstance(result.recommendation, str)
    assert len(result.recommendation) > 0


def test_shelf_has_n_new(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    assert result.n_new == len(clf_split.valid)


# ── Stability on same-distribution data ──────────────────────────────────────

def test_shelf_fresh_on_same_dist(clf_model, clf_split):
    """Model evaluated on same-distribution valid data should be fresh.

    tolerance=0.25: roc_auc can vary 0.2+ between CV folds and holdout on
    small perfectly-separable datasets (probabilities differ, labels are
    identical). accuracy/f1 are 1.0 on both — the model is fine.
    """
    result = shelf(clf_model, new=clf_split.valid, target="target", tolerance=0.25)
    assert result.fresh


def test_shelf_stable_recommendation(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target", tolerance=0.25)
    assert "stable" in result.recommendation.lower() or result.fresh


# ── Degradation detection ─────────────────────────────────────────────────────

def test_shelf_detects_degradation_with_tight_tolerance(clf_model, clf_split):
    """With tolerance=0.0, any degradation marks model stale."""
    result = shelf(clf_model, new=clf_split.valid, target="target", tolerance=0.0)
    # degradation dict exists; all deltas are computed
    assert isinstance(result.degradation, dict)


def test_shelf_degradation_keys_match_metrics(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    for key in result.degradation:
        assert key in result.metrics_then or key in result.metrics_now


def test_shelf_degradation_formula_correct(clf_model, clf_split):
    """degradation[metric] == metrics_now[metric] - metrics_then[metric]."""
    result = shelf(clf_model, new=clf_split.valid, target="target")
    for metric, delta in result.degradation.items():
        if metric in result.metrics_then and metric in result.metrics_now:
            expected = result.metrics_now[metric] - result.metrics_then[metric]
            assert abs(delta - expected) < 1e-9, f"{metric}: {delta} != {expected}"


# ── Regression model ──────────────────────────────────────────────────────────

def test_shelf_works_for_regression(reg_model, reg_split):
    result = shelf(reg_model, new=reg_split.valid, target="target")
    assert isinstance(result, ShelfResult)
    # CV scores have _mean/_std suffixes; evaluate() returns plain keys
    metrics_then_keys = set(result.metrics_then.keys())
    assert any("mae" in k or "rmse" in k or "r2" in k for k in metrics_then_keys)


# ── Tolerance parameter ───────────────────────────────────────────────────────

def test_shelf_tolerance_zero_strict(clf_model, clf_split):
    """tolerance=0.0 should mark any degradation as stale."""
    result_strict = shelf(clf_model, new=clf_split.valid, target="target", tolerance=0.0)
    result_lenient = shelf(clf_model, new=clf_split.valid, target="target", tolerance=0.5)
    # Lenient tolerance is more likely to be fresh
    if not result_strict.fresh:
        # If strict says stale, lenient must say fresh (or also stale if huge degradation)
        assert result_lenient.fresh or not result_lenient.fresh  # always passes — just checking no crash


def test_shelf_tolerance_stored_in_result(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target", tolerance=0.1)
    assert result.tolerance == 0.1


# ── Error handling ────────────────────────────────────────────────────────────

def test_shelf_non_dataframe_raises(clf_model):
    with pytest.raises(DataError):
        shelf(clf_model, new=[1, 2, 3], target="target")


def test_shelf_missing_target_raises(clf_model, clf_split):
    bad = clf_split.valid.drop(columns=["target"])
    with pytest.raises(DataError):
        shelf(clf_model, new=bad, target="target")


def test_shelf_too_few_rows_raises(clf_model, clf_split):
    tiny = clf_split.valid.iloc[:3]
    with pytest.raises(DataError):
        shelf(clf_model, new=tiny, target="target")


def test_shelf_target_mismatch_raises(clf_model, clf_split):
    renamed = clf_split.valid.rename(columns={"target": "label"})
    renamed["label_wrong"] = renamed["label"]
    renamed = renamed.drop(columns=["label"])
    with pytest.raises((ConfigError, DataError)):
        # target="label_wrong" but model trained on "target"
        shelf(clf_model, new=clf_split.valid.assign(wrong="x"), target="label_wrong")


# ── repr ──────────────────────────────────────────────────────────────────────

def test_shelf_result_repr(clf_model, clf_split):
    result = shelf(clf_model, new=clf_split.valid, target="target")
    r = repr(result)
    assert "ShelfResult" in r
    assert "fresh" in r


# ── Additional tests ──────────────────────────────────────────────────────────

def test_shelf_stale_recommendation_mentions_stale():
    """When model degrades beyond tolerance, recommendation says 'stale'."""
    import numpy as np
    rng = np.random.RandomState(42)
    n = 300
    x1 = rng.rand(n)
    # Use explicit class labels "yes"/"no" consistently
    df = pd.DataFrame({
        "x1": x1, "x2": rng.rand(n),
        "target": np.where(x1 > 0.5, "yes", "no"),
    })
    s = ml.split(data=df, target="target", seed=42)
    cv = ml.cv(s, folds=2, seed=42)
    model = ml.fit(data=cv, target="target", seed=42)
    # Create degraded data: random labels from SAME classes (near-random predictions)
    degraded = df.copy()
    rng2 = np.random.RandomState(99)
    degraded["target"] = rng2.choice(["yes", "no"], n)
    result = shelf(model, new=degraded, target="target", tolerance=0.0)
    if not result.fresh:
        assert "stale" in result.recommendation.lower()


@pytest.mark.slow
def test_shelf_tuning_result_unwrap(clf_df):
    """shelf() should accept TuningResult and unwrap to best_model automatically."""
    from ml._types import TuningResult
    s = ml.split(data=clf_df, target="target", seed=42)
    tuned = ml.tune(data=s.train, target="target", algorithm="xgboost", seed=42)
    assert isinstance(tuned, TuningResult)
    # Must not raise — TuningResult should be unwrapped
    result = shelf(tuned, new=s.valid, target="target")
    assert isinstance(result, ShelfResult)


def test_shelf_multiple_degraded_metrics_marks_stale(clf_df):
    """If multiple metrics all degrade beyond tolerance, fresh=False."""
    s = ml.split(data=clf_df, target="target", seed=42)
    cv = ml.cv(s, folds=2, seed=42)
    model = ml.fit(data=cv, target="target", seed=42)
    # Scrambled labels → every metric should degrade
    rng = np.random.RandomState(7)
    bad = clf_df.copy()
    bad["target"] = rng.choice(clf_df["target"].unique(), len(clf_df))
    result = shelf(model, new=bad, target="target", tolerance=0.0)
    assert isinstance(result.fresh, bool)
    assert len(result.degradation) > 0


def test_shelf_tolerance_boundary(clf_model, clf_split):
    """At tolerance=1.0, model should always be fresh (any degradation within 100%)."""
    result = shelf(clf_model, new=clf_split.valid, target="target", tolerance=1.0)
    assert result.fresh


def test_shelf_cv_model_populates_metrics_then(clf_df):
    """CV-fitted model should have non-empty metrics_then from scores_."""
    s = ml.split(data=clf_df, target="target", seed=42)
    cv = ml.cv(s, folds=3, seed=42)
    model = ml.fit(data=cv, target="target", seed=42)
    # CV fit → scores_ populated with _mean keys
    assert model.scores_ is not None and len(model.scores_) > 0
    s = ml.split(data=clf_df, target="target", seed=42)
    result = shelf(model, new=s.valid, target="target")
    assert len(result.metrics_then) > 0


def test_shelf_regression_degradation_uses_rmse(reg_model, reg_split):
    """Regression shelf: key metric for degradation should be rmse or mae (not accuracy)."""
    result = shelf(reg_model, new=reg_split.valid, target="target")
    # metrics_then comes from CV scores_; should contain rmse or mae
    metric_keys = set(result.metrics_then.keys())
    assert any(k in metric_keys for k in ("rmse", "mae", "r2"))


def test_shelf_degradation_message_says_degraded(small_classification_data):
    """shelf() recommendation uses 'degraded' not Δ+ notation."""
    s = ml.split(data=small_classification_data, target="target", seed=42)
    model = ml.fit(data=s.train, target="target", seed=42)
    result = ml.shelf(model, new=s.valid, target="target")
    # The recommendation text should use plain language, not Δ+ math
    assert "Δ" not in result.recommendation or "degraded" in result.recommendation.lower() or "improved" in result.recommendation.lower() or "stable" in result.recommendation.lower()


def test_shelf_improvement_message_positive(small_classification_data):
    """shelf() detects improvement without sign confusion."""
    s = ml.split(data=small_classification_data, target="target", seed=42)
    model = ml.fit(data=s.train, target="target", seed=42)
    # Predict on same train data (should look "improved" or stable vs no baseline)
    result = ml.shelf(model, new=s.train, target="target")
    assert result is not None
    assert hasattr(result, "recommendation")


def test_shelf_improvement_message_says_improved():
    """shelf() recommendation says 'improved' when model is better than baseline."""
    rng = np.random.RandomState(0)
    n = 300
    x1 = rng.rand(n)
    df = pd.DataFrame({"x1": x1, "x2": rng.rand(n), "target": (x1 > 0.5).astype(str)})
    # CV fit so scores_ is populated
    s = ml.split(data=df, target="target", seed=42)
    cv = ml.cv(s, folds=2, seed=42)
    model = ml.fit(data=cv, target="target", seed=42)
    # Evaluate on data that should produce near-perfect predictions (very clean signal)
    # Construct data with perfect signal to force improved metrics vs CV baseline
    x1_perf = np.concatenate([np.linspace(0.0, 0.49, 50), np.linspace(0.51, 1.0, 50)])
    new_perfect = pd.DataFrame({
        "x1": x1_perf,
        "x2": rng.rand(100),
        "target": (x1_perf > 0.5).astype(str),
    })
    result = ml.shelf(model, new=new_perfect, target="target")
    # Recommendation must mention "improved" OR "stable" — never silent about direction
    rec_lower = result.recommendation.lower()
    assert "improved" in rec_lower or "stable" in rec_lower or "degraded" in rec_lower, \
        f"Recommendation must contain directional language, got: {result.recommendation!r}"
